### PR TITLE
Fix unreported crash opening polymorphic relations dialog for editing

### PR DESCRIPTION
**Pull request by domi4484**
Fix unreported crash opening polymorphic relations dialog for editing.

### CODE REVIEW

File: src/app/qgsrelationaddpolymorphicdialog.cpp [hunk 1]
```diff
@@ -94,18 +90,19 @@ void QgsRelationAddPolymorphicDialog::setPolymorphicRelation( const QgsPolymorph
...
+  for ( const QString &layerId : layerIds )
+    mReferencedLayersComboBox->setItemCheckState( mReferencedLayersComboBox->findData( layerId ), Qt::Checked );
+  referencedLayersChanged();
+
   int index = 0;
   const QList<QgsRelation::FieldPair> fieldPairs = polyRel.fieldPairs();
   for ( const QgsRelation::FieldPair &fieldPair : fieldPairs )
   {
-    static_cast<QLineEdit *>( mFieldsMappingTable->cellWidget( index, 0 ) )->setText( fieldPair.referencedField() );
+    static_cast<QComboBox *>( mFieldsMappingTable->cellWidget( index, 0 ) )->setCurrentText( fieldPair.referencedField() );
```
Comment (rouault) [inline src/app/qgsrelationaddpolymorphicdialog.cpp:102]: For such down casts, the intent is better captured with:
```suggestion
    qgis::down_cast<QComboBox *>( mFieldsMappingTable->cellWidget( index, 0 ) )->setCurrentText( fieldPair.referencedField() );
```
Similarly for line below

File: src/app/qgsrelationaddpolymorphicdialog.cpp [hunk 2]
```diff
@@ -239,7 +236,7 @@ bool QgsRelationAddPolymorphicDialog::isDefinitionValid()
 
   for ( int i = 0, l = mFieldsMappingTable->rowCount(); i < l; i++ )
   {
-    isValid &= !static_cast<QLineEdit *>( mFieldsMappingTable->cellWidget( i, 0 ) )->text().isNull();
+    isValid &= static_cast<QComboBox *>( mFieldsMappingTable->cellWidget( i, 0 ) )->currentData().toInt() != -1;
```
Comment (rouault) [inline src/app/qgsrelationaddpolymorphicdialog.cpp:239]:
```suggestion
    isValid &= qgis::down_cast<QComboBox *>( mFieldsMappingTable->cellWidget( i, 0 ) )->currentData().toInt() != -1;
```
Same for below line

Comment (domi4484) [reply src/app/qgsrelationaddpolymorphicdialog.cpp]: Thank you @rouault , i was not aware of this! I have applied the changes in last commit

**Review by m-kuhn — APPROVED** [review]

File: src/app/qgsrelationaddpolymorphicdialog.cpp [hunk 3]
```diff
@@ -94,18 +90,19 @@ void QgsRelationAddPolymorphicDialog::setPolymorphicRelation( const QgsPolymorph
...
+    mReferencedLayersComboBox->setItemCheckState( mReferencedLayersComboBox->findData( layerId ), Qt::Checked );
+  referencedLayersChanged();
+
   int index = 0;
   const QList<QgsRelation::FieldPair> fieldPairs = polyRel.fieldPairs();
   for ( const QgsRelation::FieldPair &fieldPair : fieldPairs )
   {
-    static_cast<QLineEdit *>( mFieldsMappingTable->cellWidget( index, 0 ) )->setText( fieldPair.referencedField() );
-    static_cast<QgsFieldComboBox *>( mFieldsMappingTable->cellWidget( index, 1 ) )->setCurrentText( fieldPair.referencingField() );
+    qgis::down_cast<QComboBox *>( mFieldsMappingTable->cellWidget( index, 0 ) )->setCurrentText( fieldPair.referencedField() );
```
Comment (3nids) [inline src/app/qgsrelationaddpolymorphicdialog.cpp:102]: shouldn't we use qobject_cast here?

**Review by 3nids — APPROVED** [review]

File: src/app/qgsrelationaddpolymorphicdialog.cpp (hunk 3 above)
Comment (m-kuhn) [reply src/app/qgsrelationaddpolymorphicdialog.cpp:102]: I think that would be preferable

**Pull request merged**